Speed up run_doctype_migration

esoergel opened the pull request:
@dannyroberts cc @gcapalbo 
This has two changes:
1. Individually re-fetch all documents with attachments in order to get the attachments.  (Previously this was getting each attachment individually)
2. Don't write the docs to an intermediate log file.  This loses the atomicity of the operation - if there's a failure during the reads, some of the writes will have already happened. Unfortunately, we estimate the size of the intermediate file to be in the hundreds of GBs, which just isn't feasible.

esoergel committed:
Switch to get all attachments at once with the doc

dannyroberts commented on corehq/doctypemigrations/bulk_migrate.py:
```diff
@@ -5,37 +5,25 @@
+        doc_json['_attachments'] = {
+            name: _encode_data(attachment_dict)
+            for name, attachment_dict in full_doc['_attachments'].items()
+        }
```
Just wanted to confirm that `db.get` is returning unencoded attachments, and that you're not double base64-encoding the attachment payloads. I know couchdb returns them base64 encode when you ask for attachments=true, but from this code it looks like you found that couchdbkit then decodes them for you? That's surprising to me because I thought jsonobject would fail if you hand it a binary blob (since JSON strings are unicode strings, not byte strings).

esoergel replied:
Hmm, looks like you might be right - I assumed `db.fetch_attachment` and `db.get(doc_id, attachments=True)` would return the same encoding.  Will test and update.

dannyroberts replied:
Ah ok, my guess is then that this is double encoding. And now that I think of it, the other code was docs directly from the view with attachments and saving the JSON as-is, which further points in that direction.

The good news is that double encoding was probably making your time slower that it should be (since base64 inflates size of a payload by 33%).

esoergel committed:
Stop writing to disk - just read and save

This loses the atomicity of the operation - if there's a failure during
the reads, some of the writes will have already happened.
Unfortunately, we estimate the size of the intermediate file to be in
the hundreds of GBs, which just isn't feasible.

esoergel commented:
This brings the time required for the migration from ~30 minutes to 13 minutes locally.

dannyroberts commented:
This looks great. And in the end I don't think we'll ever regret that we aren't writing to an intermediate file anymore. I know I was hesitant, but talking through it, I think it was really useful in the beginning when there were more bugs, and as we've done this more and more times, I think it's less useful anyway.
